salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+        log.debug("Sent payload to rabbitmq publish daemon.")
+
+
+class AsyncReqMessageClientPool(salt.transport.MessageClientPool):
+    """
+    Wrapper class of AsyncReqMessageClientPool to avoid blocking waiting while writing data to socket.
+    """
+
+    def __init__(self, opts, args=None, kwargs=None):
+        self._closing = False
+        super().__init__(AsyncReqMessageClient, opts, args=args, kwargs=kwargs)
+
+    def close(self):
+        if self._closing:
+            return
+
+        self._closing = True
+        for message_client in self.message_clients:
+            message_client.close()
+        self.message_clients = []
+
+    def send(self, *args, **kwargs):
+        message_clients = sorted(self.message_clients, key=lambda x: len(x.send_queue))
+        return message_clients[0].send(*args, **kwargs)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+
+# TODO: unit tests!
+class AsyncReqMessageClient:
+    """
+    This class gives a future-based
+    interface to sending and receiving messages. This works around the primary
+    limitation of serialized send/recv on the underlying socket by queueing the
+    message sends in this class. In the future if we decide to attempt to multiplex
+    we can manage a pool of REQ/REP sockets-- but for now we'll just do them in serial
+
+    TODO: this implementation was adapted from ZeroMQ. Simplify this for RMQ.
+    """
+
+    def __init__(self, opts, addr, io_loop=None):
+        """
+        Create an asynchronous message client
+
+        :param dict opts: The salt opts dictionary
+        :param str addr: The interface IP address to bind to
+        :param IOLoop io_loop: A Tornado IOLoop event scheduler [tornado.ioloop.IOLoop]
+        """
+        self.opts = opts
+        self.addr = addr
+        self._rmq_connection_wrapper = RMQBlockingConnectionWrapper(opts)
+
+        if io_loop is None:
+            self.io_loop = salt.ext.tornado.ioloop.IOLoop.current()
+        else:
+            self.io_loop = io_loop
+
+        self.serial = salt.payload.Serial(self.opts)
+
+        self.send_queue = []
+        # mapping of message -> future
+        self.send_future_map = {}
+
+        self.send_timeout_map = {}  # message -> timeout
+        self._closing = False
+
+    # TODO: timeout all in-flight sessions, or error
+    def close(self):
+        try:
+            if self._closing:
+                return
+        except AttributeError:
+            # We must have been called from __del__
+            # The python interpreter has nuked most attributes already
+            return
+        else:
+            self._closing = True
+
+    # pylint: disable=W1701
+    def __del__(self):
+        self.close()
+
+    # pylint: enable=W1701
+
+    @salt.ext.tornado.gen.coroutine
+    def _internal_send_recv(self):
+        while len(self.send_queue) > 0:
```

Comment:
**suggestion** no need to check `>0`, at least, not if it's a list. Empty things in Python are considered False, so `while self.send_queue:` should work just fine here.

Reply:
Implementation carried over from zezromq.py. I'll fix it.